src/main/java/io/jenkins/plugins/checks/github/GitHubChecksContext.java

```diff
@@ -79,75 +75,64 @@ public String getURL() {
         }
 
-        return credentials;
+        return foundCredentials.get();
     }
 
     private String resolveHeadSha() {
-        if (headSha == null) {
-            Optional<SCMHead> head = scmFacade.findHead(job);
-            if (!head.isPresent()) {
-                throw new IllegalStateException("No SCM head available for job: " + job.getName());
-            }
-
-            Optional<SCMRevision> revision = scmFacade.findRevision(resolveSource(), head.get());
-            if (!revision.isPresent()) {
-                throw new IllegalStateException(
-                        String.format("No SCM revision available for repository: %s and head: %s",
-                                getRepository(), head.get().getName()));
-            }
-
-            if (revision.get() instanceof AbstractGitSCMSource.SCMRevisionImpl) {
-                headSha = ((AbstractGitSCMSource.SCMRevisionImpl) revision.get()).getHash();
-            }
-            else if (revision.get() instanceof PullRequestSCMRevision) {
-                headSha = ((PullRequestSCMRevision) revision.get()).getPullHash();
-            }
-            else {
-                throw new IllegalStateException("Unsupported revision type: " + revision.get().getClass().getName());
-            }
+        Optional<SCMHead> head = scmFacade.findHead(job);
+        if (!head.isPresent()) {
+            throw new IllegalStateException("No SCM head available for job: " + job.getName());
+        }
+
+        Optional<SCMRevision> revision = scmFacade.findRevision(resolveSource(), head.get());
+        if (!revision.isPresent()) {
+            throw new IllegalStateException(
+                    String.format("No SCM revision available for repository: %s and head: %s",
+                            getRepository(), head.get().getName()));
         }
 
-        return headSha;
+        if (revision.get() instanceof AbstractGitSCMSource.SCMRevisionImpl) {
+            return  ((AbstractGitSCMSource.SCMRevisionImpl) revision.get()).getHash();
+        }
+        else if (revision.get() instanceof PullRequestSCMRevision) {
+            return  ((PullRequestSCMRevision) revision.get()).getPullHash();
+        }
+        else {
+            throw new IllegalStateException("Unsupported revision type: " + revision.get().getClass().getName());
+        }
     }
 
     private String resolveURL(final JenkinsFacade jenkinsFacade) {
-        if (url == null) {
-            Run<?, ?> lastBuild = job.getLastBuild();
-            if (lastBuild.isLogUpdated()) {
-                url = jenkinsFacade.getAbsoluteUrl(lastBuild.getUrl());
-            }
-            else {
-                String[] tokens = lastBuild.getUrl().split("/");
-                tokens[tokens.length - 1] = String.valueOf(job.getNextBuildNumber());
-                url = jenkinsFacade.getAbsoluteUrl(tokens);
-            }
+        Run<?, ?> lastBuild = job.getLastBuild();
+        if (lastBuild.isLogUpdated()) {
```

Comment:
This doesn’t seem like a safe way to get the log is this how GitHub branch source does it?

Reply:
This is a somewhat magic code because I can't find another way to distinguish a job's stages from each other (onEntersQueue, onStarted, and onCompleted)...

say we are building #200:
* when `onEntersQueue` (not in queue yet), the last build  (seems I need to add null check) is the #199 build, so `lastBuild.isLogUpdated()` returns false.
* when `onStarted` and `onCompleted`, the last build is the #200 build and it's building, so `lastBuild.isLogUpdated()` returns true.

Reply:
but the run is the instance of the job if you work from the run won't you have all that information available to you?

Reply:
This doesn't seem right,

Take a look at how github-branch-source does it:
https://github.com/jenkinsci/github-branch-source-plugin/blob/6ba91b94ba729f69ca9e092f0aa874f92102ed67/src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubBuildStatusNotification.java#L182

the URL is retrieved from the run:
https://github.com/jenkinsci/github-branch-source-plugin/blob/42a2234b2a501f299f10206bad81c2ff9a87cd8d/src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubNotificationContext.java#L170

falling back to the job if it's not available yet.

This seems really hacky and not a good approach, imo.

Reply:
We can decide whether the job is still building by calling `isBuilding` and retrieve the URL.

But we can't decide whether the job is entering the queue or has been completed since both returns false calling `isBuilding` and false calling `isInQueue`. Therefore, we even don't know when to fall back to job when to retrieve from run...

Another way I come up now to distinguish the two state (entering the queue or on completed) is by comparing the start time of the last build and now:

        if (job.getLastBuild().getStartTimeInMillis() < Instant.now().toEpochMilli()) {
            // this build still in queue
            DisplayURLProvider.get().getJobURL(job);
        } else {
            // this build has been started
            DisplayURLProvider.get().getRunURL(job.getLastBuild());
        }

Reply:
why do we care if the build is building?

from what I can tell we should check if the run is null and if it's null we should use the jobs url.

this seems far more complicated than needed and will probably break if multiple builds are triggered around the same time.

Reply:
but there is no run at all, what consumers provide here is just the job.

Reply:
the last build will not be null unless you are building for the first time.

Reply:
consumers should be providing a run though?
except in the instance of it being in the queue which then we get the job?

Reply:
so providing one `createPublisher` with `Run` and another one with `Job`? 
 
    protected Optional<ChecksPublisher> createPublisher(final Run<?, ?> run)
    protected Optional<ChecksPublisher> createPublisher(final Job<?, ?> job)

then I'll change https://github.com/jenkinsci/checks-api-plugin/pull/4 later

Reply:
yes I would do that, and document appropriately, as far as I can tell the only consumer of the job one should be someone interested in listening to the queue which is a very specific case. the rest should operate on a run.